target/snitch_cluster/sw/apps/snax-mac-simple/untiled/src/untiled.c

```diff
@@ -30,61 +26,36 @@ int main() {
     // Wait until DMA transfer is done
     snrt_cluster_hw_barrier();
 
-    // Read the mcycle CSR (this is our way to mark/delimit a specific
-    // code region for benchmarking)
-    uint32_t pre_is_compute_core = snrt_mcycle();
-
     if (snrt_is_compute_core()) {
-        // This marks the start of the accelerator style of MAC operation
-        uint32_t csr_set = snrt_mcycle();
-
-        // Set addresses
-        write_csr(0x3d0, (uint32_t)local_a);
-        write_csr(0x3d1, (uint32_t)local_b);
-        write_csr(0x3d3, (uint32_t)local_o);
-
-        // Set configs
-        write_csr(0x3d4, 1);        // Number of iterations
-        write_csr(0x3d5, VEC_LEN);  // Vector length
-        write_csr(0x3d6, 1);        // Set simple multiplication
-
-        // Write start CSR to launch accelerator
-        write_csr(0x3c0, 0);
-
-        // Start of CSR start and poll until accelerator finishes
-        uint32_t mac_start = snrt_mcycle();
-
-        uint32_t break_poll;
-
-        while (1) {
-            // 0x3c3 is the CSR address for accelerator status
-            break_poll = read_csr(0x3c3);
-            if (break_poll == 0) {
-                break;
-            };
-        };
-
-        uint32_t mac_end = snrt_mcycle();
-        uint32_t cpu_checker;
-
-        for (uint32_t i = 0; i < (uint32_t)VEC_LEN; i++) {
-            // Check if output is same as golden output
-            if (*(local_o + i) != OUT[i]) {
-                err++;
-            };
-
-            // Compute using CPU multiplier
-            // Not the MAC output
-            cpu_checker = (*(local_a + i)) * (*(local_b + i));
+        snax_mac_setup_simple_mult(local_a, local_b, local_o, VEC_LEN);
+        snax_mac_launch();
+        snax_mac_sw_barrier();
+    }
 
-            // Compare if MAC output is same as CPU multiplier
-            if (*(local_o + i) != cpu_checker) {
-                err++;
-            };
-        };
+    // Use data mover to send output from TCDM to L3
+    if (snrt_is_dm_core()) {
+        size_t vector_size = VEC_LEN * sizeof(uint32_t);
+        snrt_dma_start_1d(OUT_TEST, local_o, vector_size);
+    }
+    snrt_cluster_hw_barrier();
+    uint32_t cycles_post = snrt_mcycle();
+    // Move tiled output data from L3 back to TCDM to check for correctness
+    if (snrt_is_dm_core()) {
+        size_t vector_size = VEC_LEN * sizeof(uint32_t);
+        snrt_dma_start_1d(local_o, OUT_TEST, vector_size);
+    }
+    // Wait until DMA transfer is done
+    snrt_cluster_hw_barrier();
 
-        uint32_t end_of_check = snrt_mcycle();
+    // Perform correctness check
+    int err = 0;
+    if (snrt_is_compute_core()) {
+        err = check_simple_mult(local_o, OUT, VEC_LEN);
+        // Compute using CPU multiplier and check
+        uint32_t cpu_output[VEC_LEN];
+        cpu_simple_mult(local_a, local_b, cpu_output, VEC_LEN);
```

Comment:
This block is a bit confusing. Could we check the results after all the computation is down? Or should we put the check result for the Mac engine in the same block of computation?

Reply:
@xiaoling-yi do you mean you want me to put the check after the `cpu_simple_mult` function?

Reply:
yes!